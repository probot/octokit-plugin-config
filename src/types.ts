export type Configuration = Record<string, unknown>;

export type API = {
  config: {
    get<T extends Configuration = Configuration>(
      options: GetOptions<T>,
    ): Promise<GetResult<T>>;
  };
};

export type GetOptions<T> = {
  /**
   * Repository owner login
   */
  owner: string;
  /**
   * Repository name
   */
  repo: string;
  /**
   * Path to configuration file
   */
  path: string;
  /**
   * Default settings that the loaded configuration will be merged into.
   *
   * An object will be merged shallowly. Pass a function for deep merges and custom merge strategies,
   * @see https://github.com/probot/octokit-plugin-config/#merging-configuration
   */
  defaults?: T;

  /**
   * Custom merge function to combine multiple configurations.
   */
  merge?: mergeFunction<T>;

  /**
   * Branch to load configuration from
   */
  branch?: string;
};

export type GetResult<T> = {
  /**
   * combined combination from all loaded files and passed `defaults` option
   */
  config: T;
  /**
   * List of files that configuration was loaded from.
   */
  files: ConfigFile[];
};

export type ConfigFile = {
  /**
   * Repository owner login
   */
  owner: string;
  /**
   * Repository name
   */
  repo: string;
  /**
   * Path to configuration file
   */
  path: string;
  /**
   * API URL for configuration file
   */
  url: string;
  /**
   * Configuration data object. Set to null if the file does not exist.
   */
  config: Configuration | null;
};

export type mergeFunction<T> = (...configs: Configuration[]) => T;
