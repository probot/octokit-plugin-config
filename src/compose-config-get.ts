import type { Octokit } from "@octokit/core";
import type {
  Configuration,
  GetOptions,
  GetResult,
  mergeFunction,
} from "./types.js";
import { getConfigFiles } from "./util/get-config-files.js";
import { deepmerge } from "@fastify/deepmerge";

const defaultMerge = deepmerge({ all: true });

/**
 * Loads configuration from one or multiple files and resolves with
 * the combined configuration as well as the list of files the configuration
 * was loaded from
 *
 * @param octokit Octokit instance
 * @param options
 */
export async function composeConfigGet<T extends Configuration = Configuration>(
  octokit: Octokit,
  {
    owner,
    repo,
    defaults = {} as T,
    merge = defaultMerge as mergeFunction<T>,
    path,
    branch,
  }: GetOptions<T>,
): Promise<GetResult<T>> {
  const files = await getConfigFiles(octokit, {
    owner,
    repo,
    path,
    branch,
  });

  const configs = files
    .map((file) => file.config)
    .reverse()
    .filter(Boolean) as Configuration[];

  return {
    files,
    config: merge(defaults, ...(configs as Configuration[])) as T,
  };
}
