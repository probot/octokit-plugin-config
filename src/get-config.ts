import { Octokit } from "@octokit/core";
import yaml from "js-yaml";

import type { Configuration } from "./types";

type GetOptions = {
  owner: string;
  repo: string;
  path: string;
};

type Config<T> = {
  config: T | null;
};

/**
 * Load configuration for a given repository and path. If the file does not exist,
 * it loads the configuration from the same file in the `.github` repository by the
 * same owner.
 *
 * @param octokit Octokit instance
 * @param options
 */
export async function getConfig<T>(
  octokit: Octokit,
  { owner, repo, path }: GetOptions
): Promise<Config<T>> {
  try {
    // https://docs.github.com/en/rest/reference/repos#get-repository-content
    const { data, headers } = await octokit.request(
      "GET /repos/{owner}/{repo}/contents/{path}",
      {
        owner,
        repo,
        path,
        mediaType: {
          format: "raw",
        },
      }
    );

    // If path is a submodule, or a folder, then a JSON string is returned with
    // the "Content-Type" header set to "application/json; charset=utf-8".
    //
    // - https://docs.github.com/en/rest/reference/repos#if-the-content-is-a-submodule
    // - https://docs.github.com/en/rest/reference/repos#if-the-content-is-a-directory
    //
    // symlinks just return the content of the linked file when requesting the raw formt,
    // so we are fine
    if (headers["content-type"] === "application/json; charset=utf-8") {
      return {
        config: null,
      };
    }

    return {
      config: ((yaml.safeLoad(data) as unknown) as T) || null,
    };
  } catch (error) {
    if (error.status === 404) {
      if (repo === ".github") {
        return {
          config: null,
        };
      }

      return getConfig(octokit, {
        owner,
        repo: ".github",
        path,
      });
    }

    throw error;
  }
}
