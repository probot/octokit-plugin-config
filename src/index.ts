import { Octokit } from "@octokit/core";

import { VERSION } from "./version.js";
import { composeConfigGet } from "./compose-config-get.js";
import type * as Types from "./types.js";
export type * from "./types.js";

/**
 * @param octokit Octokit instance
 */
export function config(octokit: Octokit): Types.API {
  return {
    config: {
      async get(options) {
        return composeConfigGet(octokit, options);
      },
    },
  };
}

config.VERSION = VERSION;

export { composeConfigGet } from "./compose-config-get.js";

export namespace createPullRequest {
  export type GetOptions<T extends Types.Configuration = Types.Configuration> =
    Types.GetOptions<T>;
  export type GetResult<T extends Types.Configuration = Types.Configuration> =
    Types.GetResult<T>;
  export type ConfigFile = Types.ConfigFile;
}
