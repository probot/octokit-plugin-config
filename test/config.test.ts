import { Octokit } from "@octokit/core";
import fetchMock from "fetch-mock";
import { config } from "../src";

const TestOctokit = Octokit.plugin(config);

const NOT_FOUND_RESPONSE = {
  status: 404,
  body: {
    message: "Not Found",
    documentation_url:
      "https://docs.github.com/rest/reference/repos#get-repository-content",
  },
};

describe("octokit.config.get", () => {
  it("README simple usage example", async () => {
    const mock = fetchMock
      .sandbox()
      .getOnce(
        "https://api.github.com/repos/octocat/hello-world/contents/.github%2Fmy-app.yml",
        `comment: 'Thank you for creating the issue!'`,
        {
          headers: {
            accept: "application/vnd.github.v3.raw",
          },
        }
      );
    const octokit = new TestOctokit({
      request: {
        fetch: mock,
      },
    });

    const { config } = await octokit.config.get({
      owner: "octocat",
      repo: "hello-world",
      filename: "my-app.yml",
    });

    expect(config).toStrictEqual({
      comment: "Thank you for creating the issue!",
    });
    expect(mock.done()).toBe(true);
  });

  it("config file missing", async () => {
    const mock = fetchMock
      .sandbox()
      .getOnce(
        "https://api.github.com/repos/octocat/hello-world/contents/.github%2Fmy-app.yml",
        NOT_FOUND_RESPONSE
      )
      .getOnce(
        "https://api.github.com/repos/octocat/.github/contents/.github%2Fmy-app.yml",
        NOT_FOUND_RESPONSE
      );

    const octokit = new TestOctokit({
      request: {
        fetch: mock,
      },
    });

    const { config } = await octokit.config.get({
      owner: "octocat",
      repo: "hello-world",
      filename: "my-app.yml",
      defaults: {
        comment: "Thank you for creating the issue!",
      },
    });

    expect(config).toStrictEqual({
      comment: "Thank you for creating the issue!",
    });
    //github.com/probot/octokit-plugin-config
    https: expect(mock.done()).toBe(true);
  });

  it("defaults option when no config present", async () => {
    const mock = fetchMock
      .sandbox()
      .getOnce(
        "https://api.github.com/repos/octocat/hello-world/contents/.github%2Fmy-app.yml",
        NOT_FOUND_RESPONSE
      )
      .getOnce(
        "https://api.github.com/repos/octocat/.github/contents/.github%2Fmy-app.yml",
        NOT_FOUND_RESPONSE
      );

    const octokit = new TestOctokit({
      request: {
        fetch: mock,
      },
    });

    const { config } = await octokit.config.get({
      owner: "octocat",
      repo: "hello-world",
      filename: "my-app.yml",
      defaults: {
        comment: "Thank you for creating the issue!",
      },
    });

    expect(config).toStrictEqual({
      comment: "Thank you for creating the issue!",
    });
    expect(mock.done()).toBe(true);
  });

  it("merges defaults option", async () => {
    const mock = fetchMock
      .sandbox()
      .getOnce(
        "https://api.github.com/repos/octocat/hello-world/contents/.github%2Fmy-app.yml",
        `config: 'value from .github/my-app.yml'`
      );

    const octokit = new TestOctokit({
      request: {
        fetch: mock,
      },
    });

    const { config } = await octokit.config.get({
      owner: "octocat",
      repo: "hello-world",
      filename: "my-app.yml",
      defaults: {
        config: "default value",
        otherConfig: "default value",
      },
    });

    expect(config).toStrictEqual({
      config: "value from .github/my-app.yml",
      otherConfig: "default value",
    });
    expect(mock.done()).toBe(true);
  });

  it("merges defaults option from .github repository", async () => {
    const mock = fetchMock
      .sandbox()
      .getOnce(
        "https://api.github.com/repos/octocat/hello-world/contents/.github%2Fmy-app.yml",
        NOT_FOUND_RESPONSE
      )
      .getOnce(
        "https://api.github.com/repos/octocat/.github/contents/.github%2Fmy-app.yml",
        `config: 'value from octocat/.github:.github/my-app.yml'`
      );

    const octokit = new TestOctokit({
      request: {
        fetch: mock,
      },
    });

    const { config } = await octokit.config.get({
      owner: "octocat",
      repo: "hello-world",
      filename: "my-app.yml",
      defaults: {
        config: "default value",
        otherConfig: "default value",
      },
    });

    expect(config).toStrictEqual({
      config: "value from octocat/.github:.github/my-app.yml",
      otherConfig: "default value",
    });
    expect(mock.done()).toBe(true);
  });
});
