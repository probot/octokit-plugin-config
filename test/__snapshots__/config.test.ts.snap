// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_extends file is missing: result 1`] = `
Object {
  "config": Object {
    "setting1": "value from repo config file",
  },
  "files": Array [
    Object {
      "config": Object {
        "setting1": "value from repo config file",
      },
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
    Object {
      "config": null,
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "base",
    },
  ],
}
`;

exports[`config file is a submodule: log 1`] = `
Object {
  "debug": [MockFunction],
  "error": [MockFunction],
  "info": [MockFunction],
  "warn": [MockFunction] {
    "calls": Array [
      Array [
        "[@probot/octokit-plugin-config] https://api.github.com/repos/octocat/hello-world/contents/.github%2Fmy-app.yml exists, but is either a directory or a submodule. Ignoring.",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;

exports[`config file is a submodule: result 1`] = `
Object {
  "config": Object {},
  "files": Array [
    Object {
      "config": null,
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
    Object {
      "config": null,
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": ".github",
    },
  ],
}
`;

exports[`config file is empty: result 1`] = `
Object {
  "config": Object {},
  "files": Array [
    Object {
      "config": Object {},
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
  ],
}
`;

exports[`octokit.config.get _extends: base with defaults and custom merge: result 1`] = `
Object {
  "config": Object {
    "otherSetting1": "value from repo config file",
    "otherSetting2": "value from base config file",
    "otherSetting3": "default value",
    "settings": Object {
      "one": "value from repo config file",
      "three": "default value",
      "two": "value from base config file",
    },
  },
  "files": Array [
    Object {
      "config": Object {
        "otherSetting1": "value from repo config file",
        "settings": Object {
          "one": "value from repo config file",
        },
      },
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
    Object {
      "config": Object {
        "otherSetting1": "value from base config file",
        "otherSetting2": "value from base config file",
        "settings": Object {
          "one": "value from base config file",
          "two": "value from base config file",
        },
      },
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "base",
    },
  ],
}
`;

exports[`octokit.config.get _extends: base: result 1`] = `
Object {
  "config": Object {
    "setting1": "value from repo config file",
    "setting2": "value from base config file",
  },
  "files": Array [
    Object {
      "config": Object {
        "setting1": "value from repo config file",
      },
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
    Object {
      "config": Object {
        "setting1": "value from base config file",
        "setting2": "value from base config file",
      },
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "base",
    },
  ],
}
`;

exports[`octokit.config.get README simple usage example: result 1`] = `
Object {
  "config": Object {
    "comment": "Thank you for creating the issue!",
  },
  "files": Array [
    Object {
      "config": Object {
        "comment": "Thank you for creating the issue!",
      },
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
  ],
}
`;

exports[`octokit.config.get config file missing: result 1`] = `
Object {
  "config": Object {
    "comment": "Thank you for creating the issue!",
  },
  "files": Array [
    Object {
      "config": null,
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
    Object {
      "config": null,
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": ".github",
    },
  ],
}
`;

exports[`octokit.config.get defaults option when no config present: result 1`] = `
Object {
  "config": Object {
    "comment": "Thank you for creating the issue!",
  },
  "files": Array [
    Object {
      "config": null,
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
    Object {
      "config": null,
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": ".github",
    },
  ],
}
`;

exports[`octokit.config.get merges deeply using defaults function: result 1`] = `
Object {
  "config": Object {
    "otherSetting1": "value from config file",
    "otherSetting2": "default value",
    "settings": Object {
      "one": "value from config file",
      "two": "default value",
    },
  },
  "files": Array [
    Object {
      "config": Object {
        "otherSetting1": "value from config file",
        "settings": Object {
          "one": "value from config file",
        },
      },
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
  ],
}
`;

exports[`octokit.config.get merges defaults option from .github repository: result 1`] = `
Object {
  "config": Object {
    "config": "value from octocat/.github:.github/my-app.yml",
    "otherConfig": "default value",
  },
  "files": Array [
    Object {
      "config": null,
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
    Object {
      "config": Object {
        "config": "value from octocat/.github:.github/my-app.yml",
      },
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": ".github",
    },
  ],
}
`;

exports[`octokit.config.get merges defaults option: result 1`] = `
Object {
  "config": Object {
    "config": "value from .github/my-app.yml",
    "otherConfig": "default value",
  },
  "files": Array [
    Object {
      "config": Object {
        "config": "value from .github/my-app.yml",
      },
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": "hello-world",
    },
  ],
}
`;

exports[`repo is .github: result 1`] = `
Object {
  "config": Object {},
  "files": Array [
    Object {
      "config": null,
      "owner": "octocat",
      "path": ".github/my-app.yml",
      "repo": ".github",
    },
  ],
}
`;
